File: Trace2JSON/Trace2JSON/Parsers/LeaksParser.m
//
//  LeaksParser.m
//  Trace2JSON
//
//  Created by Itay Brenner on 6/20/20.
//  Copyright © 2020 Itaysoft. All rights reserved.
//

#import "LeaksParser.h"
#import "InstrumentsPrivateHeaders.h"

@implementation LeaksParser

- (NSString *) instrumentName {
    return @"Leaks";
}

- (NSDictionary *) parseContext:(NSArray<XRContext *> *) contexts withRun:(XRRun *)run {
    NSMutableArray *result = [NSMutableArray new];
    
    XRLeaksRun *leaksRun = (XRLeaksRun *)run;
    for (XRLeak *leak in leaksRun.allLeaks) {
        DVT_VMUClassInfo *dvt = Ivar(leak, _layout);
        
        NSDictionary *parsedLeak = @{
            @"name": leak.name != nil ? leak.name : @"",
            @"description": dvt != nil && dvt.description ? dvt.description : @"",
            @"size": @(leak.size),
            @"count": @(leak.count),
            @"isCycle": @(leak.inCycle),
            @"isRootLeak": @(leak.isRootLeak),
            @"allocationTimestamp": @(leak.allocationTimestamp),
            @"displayAddress": leak.displayAddress != nil ? leak.displayAddress : @"",
            @"debugDescription": dvt != nil && dvt.debugDescription ? dvt.debugDescription : @"",
        };
        
        [result addObject:parsedLeak];
    }
    
    NSDictionary *returnDictionary = @{
        @"type": [self instrumentName],
        @"result": result
    };
    return returnDictionary;
}

@end

File: Trace2JSON/Parsers/LeaksParser/LeaksParser.swift
//
//  LeaksParser.swift
//  Trace2JSON
//
//  Created by Itay Brenner on 6/22/20.
//  Copyright © 2020 Itaysoft. All rights reserved.
//

import Foundation

class LeaksParser: ParserProtocol {
    private let helper = LeaksParserHelper()
    
    func parseContext(_ contexts: [XRContext], run: XRRun) -> InstrumentRun {
        let result = LeakRun()
        result.run = Int(run.runNumber())
        result.runName = run.displayName()
        
        // Workaround because some classes are inside Instrument's plugins and not it's libraries
        result.result = helper.getLeaksFrom(run)
        
        return result
    }
}
